diff -purN live.2013.12.05/BasicUsageEnvironment/BasicTaskScheduler.cpp live/BasicUsageEnvironment/BasicTaskScheduler.cpp
--- live.2013.12.05/BasicUsageEnvironment/BasicTaskScheduler.cpp	2013-12-05 22:35:42.000000000 +0000
+++ live/BasicUsageEnvironment/BasicTaskScheduler.cpp	2014-05-19 12:23:36.006112169 +0100
@@ -41,6 +41,15 @@ BasicTaskScheduler::BasicTaskScheduler(u
   if (maxSchedulerGranularity > 0) schedulerTickTask(); // ensures that we handle events frequently
 }
 
+BasicTaskScheduler::BasicTaskScheduler()
+  : fMaxSchedulerGranularity(10), fMaxNumSockets(0) {
+  FD_ZERO(&fReadSet);
+  FD_ZERO(&fWriteSet);
+  FD_ZERO(&fExceptionSet);
+
+  if (fMaxSchedulerGranularity > 0) schedulerTickTask(); // ensures that we handle events frequently
+}
+
 BasicTaskScheduler::~BasicTaskScheduler() {
 }
 
diff -purN live.2013.12.05/BasicUsageEnvironment/include/BasicUsageEnvironment.hh live/BasicUsageEnvironment/include/BasicUsageEnvironment.hh
--- live.2013.12.05/BasicUsageEnvironment/include/BasicUsageEnvironment.hh	2013-12-05 22:35:42.000000000 +0000
+++ live/BasicUsageEnvironment/include/BasicUsageEnvironment.hh	2014-05-19 12:22:50.260960874 +0100
@@ -54,8 +54,8 @@ public:
   virtual ~BasicTaskScheduler();
 
 protected:
-  BasicTaskScheduler(unsigned maxSchedulerGranularity);
-      // called only by "createNew()"
+  BasicTaskScheduler(unsigned maxSchedulerGranularity); // called only by "createNew()"
+  BasicTaskScheduler(); // called by "bscheduler.cpp"
 
   static void schedulerTickTask(void* clientData);
   void schedulerTickTask();
diff -purN live.2013.12.05/BasicUsageEnvironment/Makefile live/BasicUsageEnvironment/Makefile
--- live.2013.12.05/BasicUsageEnvironment/Makefile	2014-05-19 13:52:09.100341042 +0100
+++ live/BasicUsageEnvironment/Makefile	2014-05-19 13:44:36.472844558 +0100
@@ -1,6 +1,6 @@
 INCLUDES = -Iinclude -I../UsageEnvironment/include -I../groupsock/include
-PREFIX = /usr/local
-LIBDIR = $(PREFIX)/lib
+
+
 ##### Change the following for your environment:
 # 'CURRENT':'REVISION':'AGE' are updated - whenever a library changes - as follows:
 # The library code changes, but without any changes to the API (i.e., interfaces) => increment REVISION
@@ -39,13 +39,12 @@ B_REFSW_CFLAGS += -D_FILE_OFFSET_BITS=64
 DESTDIR = ../../../../../../nexus/bin
 # LIBDIR is used if you want the libraries to be in a subdirectory ... like /usr/local/lib
 LIBDIR = 
-PREFIX = 
 
 B_REFSW_ARCH ?= mipsel-uclibc
-ifeq ($(filter ${B_REFSW_ARCH}, mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
-CROSS_COMPILE ?= mips-linux-uclibc-
+ifeq ($(filter ${B_REFSW_ARCH}, arm-linux mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
+CROSS_COMPILE ?= $(B_REFSW_ARCH)-
 else
-CROSS_COMPILE ?= mipsel-linux-uclibc-
+CROSS_COMPILE ?= mipsel-linux-
 endif
 
 COMPILE_OPTS =		$(INCLUDES) -I. -DSOCKLEN_T=socklen_t -DNO_STRSTREAM=1 -DNO_SSTREAM=1 $(B_REFSW_CFLAGS) -O2 -DXLOCALE_NOT_USED=1 -fPIC
@@ -66,7 +65,8 @@ LIB_SUFFIX =	 	$(SHORT_LIB_SUFFIX).$($(N
 LIBRARY_LINK_OPTS =	-shared $(LDFLAGS) -nostdlib -Wl,--no-undefined,--gc-sections -L$(ANDROID_LIB) \
                     $(ANDROID_LIB)/crtbegin_so.o $(ANDROID_LIB)/crtend_so.o \
                     -Wl,--no-whole-archive -Wl,-dy,-ldl,-lstdc++,-lcutils,-lc,-lm \
-                    -Wl,--whole-archive -Wl,-lgcc,-lsupc++
+                    -Wl,--whole-archive -Wl,-lgcc,-lsupc++ \
+                    -L$(DESTDIR) -Wl,-lUsageEnvironment
 
 LIBS_FOR_CONSOLE_APPLICATION =
 LIBS_FOR_GUI_APPLICATION =
@@ -111,7 +111,7 @@ install1: libBasicUsageEnvironment.$(LIB
 	  install -m 644 include/*.hh $(DESTDIR)$(PREFIX)/include/BasicUsageEnvironment
 	  install -m 644 libBasicUsageEnvironment.$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)
 install_shared_libraries: libBasicUsageEnvironment.$(LIB_SUFFIX)
-	  ln -s $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).$(SHORT_LIB_SUFFIX)
-	  ln -s $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).so
+	  ln -fs $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).$(SHORT_LIB_SUFFIX)
+	  ln -fs $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).so
 
 ##### Any additional, platform-specific rules come here:
diff -purN live.2013.12.05/groupsock/GroupsockHelper.cpp live/groupsock/GroupsockHelper.cpp
--- live.2013.12.05/groupsock/GroupsockHelper.cpp	2013-12-05 22:35:43.000000000 +0000
+++ live/groupsock/GroupsockHelper.cpp	2014-05-19 13:36:16.968187061 +0100
@@ -30,6 +30,175 @@ extern "C" int initializeWinsockIfNecess
 #define initializeWinsockIfNecessary() 1
 #endif
 #include <stdio.h>
+#include <pthread.h>
+
+/*
+ * This function accepts as input a socket for which a peer socket address must be determined for it.
+ * Then the address is converted into a string and returned. If an error occurs, NULL is returned.
+ */
+char * peer_addr(int s, char *buf, size_t bufsiz) 
+{
+    int rc; /* Status return code */
+    struct sockaddr_in adr_inet;/* AF_INET */
+    socklen_t len_inet; /* length */
+
+    /*
+     * Obtain the address of the socket:
+     */
+    len_inet = sizeof adr_inet;
+
+    // int getpeername(int socket, struct sockaddr *restrict address, socklen_t *restrict address_len);
+    rc = getpeername(s, (struct sockaddr *)&adr_inet, &len_inet);
+    if ( rc == -1) {
+       return NULL; /* Failed */
+    }
+
+    /*
+     * Convert address into a string form that can be displayed:
+     */
+    //fprintf(stderr,"buf size (%u); addr (%s); port (%u)\n", bufsiz, inet_ntoa(adr_inet.sin_addr), (unsigned)ntohs(adr_inet.sin_port) );
+    rc = snprintf(buf, bufsiz, "%s:%u", inet_ntoa(adr_inet.sin_addr), (unsigned)ntohs(adr_inet.sin_port) );
+
+    if ( rc == -1 )
+    {
+       return NULL; /* Buffer too small */
+    }
+    return buf;
+}
+
+int SocketOptions ( int fd )
+{
+    int type;
+    socklen_t length = sizeof( int );
+    char buf[32];
+
+    if (fd>0)
+    {
+        // SO_SNDBUF, SO_RCVBUF, SO_TYPE,
+        getsockopt( fd, SOL_SOCKET, SO_TYPE, &type, &length );
+
+#ifdef DEBUG
+        fprintf(stderr, "SO_TYPE is (%s); ", (type==SOCK_STREAM)?"TCP":(type==SOCK_DGRAM)?"DGRAM":"RAW" );
+#endif
+
+        if (type==SOCK_STREAM)
+        {
+#if 0
+            struct sockaddr address;
+            int address_len=sizeof(address);
+            int rc = getpeername(fd, &address, &address_len );
+            if (rc==0)
+            {
+            }
+            else
+            {
+                fprintf(stderr, "getpeername error (%u) ... %s\n", errno, strerror(errno) );
+            }
+#else
+            if (peer_addr ( fd, buf, sizeof(buf) )==NULL)
+            {
+                fprintf(stderr, "peer_addr error (%u) ... %s\n", errno, strerror(errno) );
+            }
+            else
+            {
+                fprintf(stderr, "addr (%s)\n", buf );
+            }
+#endif
+        }
+        else
+        {
+            fprintf(stderr, "no addr; \n" );
+        }
+
+        // SocketMacAddress (fd);
+    }
+
+    return 0;
+}
+
+/* ******************************************************************************************************
+*                                                                                                       *
+*   PrintAddrPort - outputs to stderr the IP address and port used in the sockaddr structure.           *
+*                                                                                                       *
+********************************************************************************************************/
+int PrintAddrPort ( sockaddr_in * dest )
+{
+    if (dest==NULL) return -1;
+
+    #ifdef DEBUG
+    fprintf(stderr,"%lx(%s:%u)", ((unsigned long int) pthread_self()&0xFFFFFFFF), inet_ntoa(dest->sin_addr), htons(dest->sin_port) );
+    #endif
+    return 0;
+}
+
+/* ******************************************************************************************************
+*                                                                                                       *
+*   DumpBuffer - dumps the specified buffer in hex and ascii                                            *
+*                                                                                                       *
+********************************************************************************************************/
+void DumpBuffer ( const char * funccaller, const unsigned char * buffer, unsigned int bufferlen)
+{
+#if DEBUG
+    unsigned short int idx2 = 0;
+    char hexline[256];
+    char asciiline[20];
+    char twochar[3];
+
+    memset(hexline, 0, sizeof(hexline) );
+    memset(asciiline, 0, 20);
+
+    for (idx2=0; idx2<bufferlen; idx2++)
+    {
+        sprintf(twochar, "%02x", buffer[idx2]);
+        twochar[2] = '\0';
+        if ( (strlen(hexline) + strlen(twochar)) < sizeof(hexline) ) {
+            strcat(hexline, twochar);
+        } else {
+            fprintf(stderr, "%s: (%s)\n", funccaller, hexline );
+            fprintf(stderr, "%s: hexline buffer len (%d) exceeded\n", __FUNCTION__, sizeof(hexline) );fflush(stdout);fflush(stderr);
+            break;
+        }
+        #if 0
+        if (isprint(buffer[idx2]))
+        {
+            asciiline[idx2%16] = (char) buffer[idx2];
+        }
+        else
+        {
+            asciiline[idx2%16] = '.';
+        }
+
+        /* output line is 16 bytes */
+        if (idx2%16 == 15)
+        {
+            printf("%s '%s'\n", hexline, asciiline );
+            memset(hexline, 0, 64);
+            memset(asciiline, 0, 20);
+        }
+        #else
+        if (idx2%2==1) {
+            strcat(hexline, " ");
+        }
+        #endif
+    }
+
+    #if 0
+    /* partial last line */
+    if (idx2%16 != 0)
+    {
+        fprintf(stderr, "%s: (%s)\n", funccaller, hexline );fflush(stdout);fflush(stderr);
+    }
+    #else
+    fprintf(stderr, "%s: (%s)\n", funccaller, hexline );fflush(stdout);fflush(stderr);
+    #endif
+#else
+    ((void)funccaller);
+    ((void)buffer); // UNUSED
+    ((void)bufferlen); // UNUSED
+#endif
+
+    return;
+}
 
 // By default, use INADDR_ANY for the sending and receiving interfaces:
 netAddressBits SendingInterfaceAddr = INADDR_ANY;
@@ -295,15 +464,34 @@ int readSocket(UsageEnvironment& env,
   } else if (bytesRead == 0) {
     // "recvfrom()" on a stream socket can return 0 if the remote end has closed the connection.  Treat this as an error:
     return -1;
+  } else {
+#if DEBUG
+    fflush(stdout);fflush(stderr);
+    fprintf(stderr,"\n%lx: recvfrom(sd %u,ip %s:%u) %d bytes (%c%c%c%c%c%c%c%c%c%c%c)\n", ((unsigned long int) pthread_self()&0xFFFFFFFF), 
+            socket, inet_ntoa(fromAddress.sin_addr), htons(fromAddress.sin_port), bytesRead, strerror(errno),
+            buffer[0], buffer[1], buffer[2], buffer[3], buffer[4], buffer[5], buffer[6], buffer[7], buffer[8],
+            buffer[9], buffer[10] );
+    fflush(stdout);fflush(stderr);
+    if (bytesRead == 56) {
+        DumpBuffer ( __FUNCTION__, (unsigned char*)buffer, 56 );
+    }
+    SocketOptions(socket);
+#endif
   }
 
   return bytesRead;
 }
 
+#ifdef DEBUG
+static struct timeval timeprev, timenow;
+#endif
+static unsigned long int pktcnt=0;
+static unsigned long int gspktcnt=0;
 Boolean writeSocket(UsageEnvironment& env,
 		    int socket, struct in_addr address, Port port,
 		    u_int8_t ttlArg,
 		    unsigned char* buffer, unsigned bufferSize) {
+	gspktcnt++;
 	do {
 		if (ttlArg != 0) {
 			// Before sending, set the socket's TTL:
@@ -321,8 +509,39 @@ Boolean writeSocket(UsageEnvironment& en
 		}
 
 		MAKE_SOCKADDR_IN(dest, address.s_addr, port.num());
+#if DEBUG
+		if(gspktcnt%1000==999) {
+			   printf("sendto skt %u:%s(0x%x):htons %u(0x%x); raw %u(0x%x); bytes %u; ", socket, inet_ntoa(address), address.s_addr, htons(port.num()), htons(port.num()), port.num(), port.num(), bufferSize );
+			   {
+				   char line[64];
+				   char nibble[3];
+				   int idx=0;
+				   unsigned char * pos= (unsigned char*) &dest;
+				   memset(line, 0, sizeof line);
+				   for (idx=0; idx<8;idx++) {
+					   sprintf(nibble,"%02x", pos[idx] );
+					   nibble[2] = '\0';
+					   strcat(line, nibble);
+					   if ((idx%4==3)) { strcat(line, " "); }
+				   }
+				   printf (("%s: dest struct (%s); ", __FUNCTION__, line ));
+			   }
+			   fflush(stdout); 
+		}
+#endif
 		int bytesSent = sendto(socket, (char*)buffer, bufferSize, 0,
 			               (struct sockaddr*)&dest, sizeof dest);
+        pktcnt++;
+#ifdef DEBUG
+        gettimeofday(&timenow, NULL);
+        if (timenow.tv_sec > (timeprev.tv_sec+5) ) {
+          timeprev = timenow;
+          fprintf(stderr, "socket %u; cnt %lu; ", socket, pktcnt);
+          PrintAddrPort ( &dest ); // 10 02 e7 6e 0a 00 00 08
+          SocketOptions(socket);
+        }
+#endif
+
 		if (bytesSent != (int)bufferSize) {
 			char tmpBuf[100];
 			sprintf(tmpBuf, "writeSocket(%d), sendTo() error: wrote %d bytes instead of %u: ", socket, bytesSent, bufferSize);
diff -purN live.2013.12.05/groupsock/Makefile live/groupsock/Makefile
--- live.2013.12.05/groupsock/Makefile	2014-05-19 13:52:09.092340946 +0100
+++ live/groupsock/Makefile	2014-05-19 13:30:22.283193142 +0100
@@ -1,6 +1,6 @@
 INCLUDES = -Iinclude -I../UsageEnvironment/include
-PREFIX = /usr/local
-LIBDIR = $(PREFIX)/lib
+
+
 ##### Change the following for your environment:
 # 'CURRENT':'REVISION':'AGE' are updated - whenever a library changes - as follows:
 # The library code changes, but without any changes to the API (i.e., interfaces) => increment REVISION
@@ -39,13 +39,12 @@ B_REFSW_CFLAGS += -D_FILE_OFFSET_BITS=64
 DESTDIR = ../../../../../../nexus/bin
 # LIBDIR is used if you want the libraries to be in a subdirectory ... like /usr/local/lib
 LIBDIR = 
-PREFIX = 
 
 B_REFSW_ARCH ?= mipsel-uclibc
-ifeq ($(filter ${B_REFSW_ARCH}, mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
-CROSS_COMPILE ?= mips-linux-uclibc-
+ifeq ($(filter ${B_REFSW_ARCH}, arm-linux mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
+CROSS_COMPILE ?= $(B_REFSW_ARCH)-
 else
-CROSS_COMPILE ?= mipsel-linux-uclibc-
+CROSS_COMPILE ?= mipsel-linux-
 endif
 
 COMPILE_OPTS =		$(INCLUDES) -I. -DSOCKLEN_T=socklen_t -DNO_STRSTREAM=1 -DNO_SSTREAM=1 $(B_REFSW_CFLAGS) -O2 -DXLOCALE_NOT_USED=1 -fPIC
@@ -66,7 +65,8 @@ LIB_SUFFIX =	 	$(SHORT_LIB_SUFFIX).$($(N
 LIBRARY_LINK_OPTS =	-shared $(LDFLAGS) -nostdlib -Wl,--no-undefined,--gc-sections -L$(ANDROID_LIB) \
                     $(ANDROID_LIB)/crtbegin_so.o $(ANDROID_LIB)/crtend_so.o \
                     -Wl,--no-whole-archive -Wl,-dy,-ldl,-lstdc++,-lcutils,-lc,-lm \
-                    -Wl,--whole-archive -Wl,-lgcc,-lsupc++
+                    -Wl,--whole-archive -Wl,-lgcc,-lsupc++ \
+                    -L$(DESTDIR) -Wl,-lBasicUsageEnvironment,-lUsageEnvironment
 
 LIBS_FOR_CONSOLE_APPLICATION =
 LIBS_FOR_GUI_APPLICATION =
@@ -113,7 +113,7 @@ install1: libgroupsock.$(LIB_SUFFIX)
 	  install -m 644 include/*.hh include/*.h $(DESTDIR)$(PREFIX)/include/groupsock
 	  install -m 644 libgroupsock.$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)
 install_shared_libraries: libgroupsock.$(LIB_SUFFIX)
-	  ln -s libgroupsock.$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/libgroupsock.$(SHORT_LIB_SUFFIX)
-	  ln -s libgroupsock.$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/libgroupsock.so
+	  ln -fs libgroupsock.$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/libgroupsock.$(SHORT_LIB_SUFFIX)
+	  ln -fs libgroupsock.$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/libgroupsock.so
 
 ##### Any additional, platform-specific rules come here:
diff -purN live.2013.12.05/liveMedia/include/OnDemandServerMediaSubsession.hh live/liveMedia/include/OnDemandServerMediaSubsession.hh
--- live.2013.12.05/liveMedia/include/OnDemandServerMediaSubsession.hh	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/include/OnDemandServerMediaSubsession.hh	2014-05-19 12:37:18.718813216 +0100
@@ -43,6 +43,7 @@ protected: // we're a virtual base class
 
 protected: // redefined virtual functions
   virtual char const* sdpLines();
+  virtual char const* sdpLines( const char * filename );
   virtual void getStreamParameters(unsigned clientSessionId,
 				   netAddressBits clientAddress,
                                    Port const& clientRTPPort,
diff -purN live.2013.12.05/liveMedia/include/PassiveServerMediaSubsession.hh live/liveMedia/include/PassiveServerMediaSubsession.hh
--- live.2013.12.05/liveMedia/include/PassiveServerMediaSubsession.hh	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/include/PassiveServerMediaSubsession.hh	2014-05-19 12:37:30.851118427 +0100
@@ -45,6 +45,7 @@ protected:
 
 protected: // redefined virtual functions
   virtual char const* sdpLines();
+  virtual char const* sdpLines( const char *);
   virtual void getStreamParameters(unsigned clientSessionId,
 				   netAddressBits clientAddress,
                                    Port const& clientRTPPort,
diff -purN live.2013.12.05/liveMedia/include/RTSPClient.hh live/liveMedia/include/RTSPClient.hh
--- live.2013.12.05/liveMedia/include/RTSPClient.hh	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/include/RTSPClient.hh	2014-05-19 12:38:00.139855236 +0100
@@ -162,6 +162,9 @@ public:
       // Parses "url" as "rtsp://[<username>[:<password>]@]<server-address-or-name>[:<port>][/<stream-name>]"
       // (Note that the returned "username" and "password" are either NULL, or heap-allocated strings that the caller must later delete[].)
 
+  // BRCM: CAD 2013-10-21
+  unsigned describeStatus() const { return fDescribeStatusCode; }
+
   void setUserAgentString(char const* userAgentName);
       // sets an alternative string to be used in RTSP "User-Agent:" headers
 
@@ -171,6 +174,15 @@ public:
 
   static unsigned responseBufferSize;
 
+  // JJ/BRCM 20061026
+  char* serverType(void) { return fServerTypeStr; }
+
+  // SSOOD/BRCM 20100303
+  char* responseMessage(void) { return fResponseMessageStr; }
+
+  // SSOOD/BRCM 20100416 
+  void setAdditionalHeaders(char *hdr) { fAdditionHeadersStr = hdr; } 
+
 public: // Some compilers complain if this is "private:"
   // The state of a request-in-progress:
   class RequestRecord {
@@ -219,7 +231,9 @@ protected:
   virtual ~RTSPClient();
 
   void reset();
+public:
   void setBaseURL(char const* url);
+protected:
   int grabSocket(); // allows a subclass to reuse our input socket, so that it won't get closed when we're deleted
   virtual unsigned sendRequest(RequestRecord* request);
   virtual Boolean setRequestFields(RequestRecord* request,
@@ -309,7 +323,11 @@ private:
   unsigned char fTCPStreamIdCount; // used for (optional) RTP/TCP
   char* fLastSessionId;
   unsigned fSessionTimeoutParameter; // optionally set in response "Session:" headers
-  char* fResponseBuffer;
+
+public:
+  // BRCM: CAD 2013-10-21
+  unsigned fDescribeStatusCode; // 0: OK; 1: connection failed; 2: stream unavailable
+private:
   unsigned fResponseBytesAlreadySeen, fResponseBufferBytesLeft;
   RequestQueue fRequestsAwaitingConnection, fRequestsAwaitingHTTPTunneling, fRequestsAwaitingResponse;
 
@@ -317,6 +335,23 @@ private:
   char fSessionCookie[33];
   unsigned fSessionCookieCounter;
   Boolean fHTTPTunnelingConnectionIsPending;
+
+  // JJ/BRCM 20061026 - save Server type for users of RTSPClient
+  char *fServerTypeStr;
+
+  // SSOOD/BRCM 20100303 - save the last response headers
+  char *fResponseMessageStr;
+
+  // SSOOD/BRCM 20100416 - save the additional headers to be included w/ outgoing messages
+  char *fAdditionHeadersStr;
+
+public:
+  // CAD/BRCM 20131218 - used to block while waiting for responses for OPTIONS, DESCRIBE, SETUP, PLAY, etc.
+  char fWatchVariableForSyncInterface;
+  char* fResponseBuffer;
+  char* fResultString;
+  int fResultCode;
+
 };
 
 
diff -purN live.2013.12.05/liveMedia/include/RTSPCommon.hh live/liveMedia/include/RTSPCommon.hh
--- live.2013.12.05/liveMedia/include/RTSPCommon.hh	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/include/RTSPCommon.hh	2014-05-19 12:38:34.416717383 +0100
@@ -25,10 +25,14 @@ along with this library; if not, write t
 #include "Boolean.hh"
 #endif
 
+#ifdef __cplusplus /* these header files should not be included for gcc compiler */
+
 #ifndef _MEDIA_HH
 #include <Media.hh> // includes some definitions perhaps needed for Borland compilers?
 #endif
 
+#endif   /* __cplusplus */
+
 #if defined(__WIN32__) || defined(_WIN32) || defined(_QNX4)
 #define _strncasecmp _strnicmp
 #define snprintf _snprintf
@@ -38,6 +42,9 @@ along with this library; if not, write t
 
 #define RTSP_PARAM_STRING_MAX 200
 
+#define DEBUG 1
+
+#ifdef __cplusplus /* these header files should not be included for gcc compiler */
 Boolean parseRTSPRequestString(char const *reqStr, unsigned reqStrSize,
 			       char *resultCmdName,
 			       unsigned resultCmdNameMaxSize,
@@ -63,5 +70,6 @@ Boolean RTSPOptionIsSupported(char const
 char const* dateHeader(); // A "Date:" header that can be used in a RTSP (or HTTP) response 
 
 void ignoreSigPipeOnSocket(int socketNum);
+#endif   /* __cplusplus */
 
 #endif
diff -purN live.2013.12.05/liveMedia/include/RTSPServer.hh live/liveMedia/include/RTSPServer.hh
--- live.2013.12.05/liveMedia/include/RTSPServer.hh	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/include/RTSPServer.hh	2014-05-19 12:39:03.941460244 +0100
@@ -21,6 +21,19 @@ along with this library; if not, write t
 #ifndef _RTSP_SERVER_HH
 #define _RTSP_SERVER_HH
 
+#include "RTSPCommon.hh"
+
+// CAD/BRCM 20121031 - Added to support RTSP in ip_streamer
+#ifdef __cplusplus /* these prototypes should not be name mangled for gcc compiler */
+extern "C" {
+#endif
+
+const char * LiveMediaGetVersion ( void );
+
+// CAD/BRCM 20121031 - Added to support RTSP in ip_streamer (want gcc to ignore g++ stuff)
+#ifdef __cplusplus
+}
+
 #ifndef _SERVER_MEDIA_SESSION_HH
 #include "ServerMediaSession.hh"
 #endif
@@ -238,12 +251,14 @@ public: // should be protected, but some
     static void continueHandlingREGISTER(ParamsForREGISTER* params);
     virtual void continueHandlingREGISTER1(ParamsForREGISTER* params);
 
+  public:
     // Shortcuts for setting up a RTSP response (prior to sending it):
     void setRTSPResponse(char const* responseStr);
     void setRTSPResponse(char const* responseStr, u_int32_t sessionId);
     void setRTSPResponse(char const* responseStr, char const* contentStr);
     void setRTSPResponse(char const* responseStr, u_int32_t sessionId, char const* contentStr);
 
+  protected:
     RTSPServer& fOurServer;
     Boolean fIsActive;
     int fClientInputSocket, fClientOutputSocket;
@@ -365,7 +380,6 @@ private:
   unsigned fReclamationTestSeconds;
 };
 
-
 ////////// A subclass of "RTSPServer" that implements the "REGISTER" command to set up proxying on the specified URL //////////
 
 class RTSPServerWithREGISTERProxying: public RTSPServer {
@@ -399,5 +413,8 @@ private:
   char* fAllowedCommandNames;
   UserAuthenticationDatabase* fAuthDBForREGISTER;
 }; 
+// CAD/BRCM 20121031 - Added to support RTSP in ip_streamer
+#endif /* __cplusplus */
+
 
 #endif
diff -purN live.2013.12.05/liveMedia/include/ServerMediaSession.hh live/liveMedia/include/ServerMediaSession.hh
--- live.2013.12.05/liveMedia/include/ServerMediaSession.hh	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/include/ServerMediaSession.hh	2014-05-19 12:39:41.914415448 +0100
@@ -24,6 +24,8 @@ along with this library; if not, write t
 #ifndef _SERVER_MEDIA_SESSION_HH
 #define _SERVER_MEDIA_SESSION_HH
 
+#include "RTSPCommon.hh"
+
 #ifndef _MEDIA_HH
 #include "Media.hh"
 #endif
@@ -56,6 +58,7 @@ public:
       // Note: The caller is responsible for freeing the returned string
 
   char const* streamName() const { return fStreamName; }
+  char const* sInfoSDPString() const { return fInfoSDPString; }
 
   Boolean addSubsession(ServerMediaSubsession* subsession);
   unsigned numSubsessions() const { return fSubsessionCounter; }
@@ -82,6 +85,7 @@ protected:
 		     char const* info, char const* description,
 		     Boolean isSSM, char const* miscSDPLines);
   // called only by "createNew()"
+  char* fInfoSDPString;
 
   virtual ~ServerMediaSession();
 
@@ -98,7 +102,6 @@ private:
   unsigned fSubsessionCounter;
 
   char* fStreamName;
-  char* fInfoSDPString;
   char* fDescriptionSDPString;
   char* fMiscSDPLines;
   struct timeval fCreationTime;
@@ -126,6 +129,7 @@ public:
   unsigned trackNumber() const { return fTrackNumber; }
   char const* trackId();
   virtual char const* sdpLines() = 0;
+  virtual char const* sdpLines(const char *) = 0;
   virtual void getStreamParameters(unsigned clientSessionId, // in
 				   netAddressBits clientAddress, // in
 				   Port const& clientRTPPort, // in
@@ -175,6 +179,7 @@ public:
   // address and port number fields in SDP descriptions need to be non-zero:
   void setServerAddressAndPortForSDP(netAddressBits addressBits,
 				     portNumBits portBits);
+  char const* infoSDPString() const;
 
 protected: // we're a virtual base class
   ServerMediaSubsession(UsageEnvironment& env);
diff -purN live.2013.12.05/liveMedia/Locale.cpp live/liveMedia/Locale.cpp
--- live.2013.12.05/liveMedia/Locale.cpp	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/Locale.cpp	2014-05-19 12:39:53.370703660 +0100
@@ -21,6 +21,7 @@ along with this library; if not, write t
 
 #include "Locale.hh"
 #include <strDup.hh>
+#include <cstdio>
 
 Locale::Locale(char const* newLocale, LocaleCategory category) {
 #ifndef LOCALE_NOT_USED
diff -purN live.2013.12.05/liveMedia/Makefile live/liveMedia/Makefile
--- live.2013.12.05/liveMedia/Makefile	2014-05-19 13:52:09.088340923 +0100
+++ live/liveMedia/Makefile	2014-05-19 13:31:16.296563172 +0100
@@ -1,6 +1,6 @@
-INCLUDES = -Iinclude -I../UsageEnvironment/include -I../groupsock/include
-PREFIX = /usr/local
-LIBDIR = $(PREFIX)/lib
+INCLUDES = -Iinclude -I../UsageEnvironment/include -I../BasicUsageEnvironment/include -I../groupsock/include
+
+
 ##### Change the following for your environment:
 # 'CURRENT':'REVISION':'AGE' are updated - whenever a library changes - as follows:
 # The library code changes, but without any changes to the API (i.e., interfaces) => increment REVISION
@@ -34,18 +34,22 @@ B_REFSW_CFLAGS += -DLINUX -DANDROID
 B_REFSW_CFLAGS += -pipe
 B_REFSW_CFLAGS += -W
 B_REFSW_CFLAGS += -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE
+B_REFSW_CFLAGS += -DRTSPCLIENT_SYNCHRONOUS_INTERFACE
+
+ifeq ($(WMS_SUPPORT), y)
+B_REFSW_CFLAGS += -DB_HAS_WMS
+endif
 
 # DESTDIR is where you want the shared object libraries to end up (relative to the build directory obj.97425/BSEAV/lib/livemedia/mipsel-uclibc.debug/live/BasicUsageEnvironment
 DESTDIR = ../../../../../../nexus/bin
 # LIBDIR is used if you want the libraries to be in a subdirectory ... like /usr/local/lib
 LIBDIR = 
-PREFIX = 
 
 B_REFSW_ARCH ?= mipsel-uclibc
-ifeq ($(filter ${B_REFSW_ARCH}, mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
-CROSS_COMPILE ?= mips-linux-uclibc-
+ifeq ($(filter ${B_REFSW_ARCH}, arm-linux mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
+CROSS_COMPILE ?= $(B_REFSW_ARCH)-
 else
-CROSS_COMPILE ?= mipsel-linux-uclibc-
+CROSS_COMPILE ?= mipsel-linux-
 endif
 
 COMPILE_OPTS =		$(INCLUDES) -I. -DSOCKLEN_T=socklen_t -DNO_STRSTREAM=1 -DNO_SSTREAM=1 $(B_REFSW_CFLAGS) -O2 -DXLOCALE_NOT_USED=1 -fPIC
@@ -66,7 +70,8 @@ LIB_SUFFIX =	 	$(SHORT_LIB_SUFFIX).$($(N
 LIBRARY_LINK_OPTS =	-shared $(LDFLAGS) -nostdlib -Wl,--no-undefined,--gc-sections -L$(ANDROID_LIB) \
                     $(ANDROID_LIB)/crtbegin_so.o $(ANDROID_LIB)/crtend_so.o \
                     -Wl,--no-whole-archive -Wl,-dy,-ldl,-lstdc++,-lcutils,-lc,-lm \
-                    -Wl,--whole-archive -Wl,-lgcc,-lsupc++
+                    -Wl,--whole-archive -Wl,-lgcc,-lsupc++ \
+                    -L$(DESTDIR) -Wl,-lgroupsock,-lBasicUsageEnvironment,-lUsageEnvironment
 
 LIBS_FOR_CONSOLE_APPLICATION =
 LIBS_FOR_GUI_APPLICATION =
@@ -436,7 +441,7 @@ install1: $(LIVEMEDIA_LIB)
 	 install -m 644 include/*.hh $(DESTDIR)$(PREFIX)/include/liveMedia
 	 install -m 644 $(LIVEMEDIA_LIB) $(DESTDIR)$(LIBDIR)
 install_shared_libraries: $(LIVEMEDIA_LIB)
-	 ln -s $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).$(SHORT_LIB_SUFFIX)
-	 ln -s $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).so
+	 ln -fs $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).$(SHORT_LIB_SUFFIX)
+	 ln -fs $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).so
 
 ##### Any additional, platform-specific rules come here:
diff -purN live.2013.12.05/liveMedia/Media.cpp live/liveMedia/Media.cpp
--- live.2013.12.05/liveMedia/Media.cpp	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/Media.cpp	2014-05-19 12:42:49.647137776 +0100
@@ -35,7 +35,8 @@ Medium::Medium(UsageEnvironment& env)
 
 Medium::~Medium() {
   // Remove any tasks that might be pending for us:
-  fEnviron.taskScheduler().unscheduleDelayedTask(fNextTask);
+  /* CAD 2014-04-08: this function crashes when you ctrl-c BIP_RtspServer derived class */
+  if(fNextTask>0) fEnviron.taskScheduler().unscheduleDelayedTask(fNextTask);
 }
 
 Boolean Medium::lookupByName(UsageEnvironment& env, char const* mediumName,
diff -purN live.2013.12.05/liveMedia/MultiFramedRTPSource.cpp live/liveMedia/MultiFramedRTPSource.cpp
--- live.2013.12.05/liveMedia/MultiFramedRTPSource.cpp	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/MultiFramedRTPSource.cpp	2014-05-19 12:43:29.000127853 +0100
@@ -23,6 +23,15 @@ along with this library; if not, write t
 #include "GroupsockHelper.hh"
 #include <string.h>
 
+/* PT: for some ASF streams (dido -> tw -> an), default of 10000  */
+/* is not enough. BJ/CH team (AY) suggested increasing this for   */
+/* TCP mode ASF over IP. Only do this for WMS case.    */
+#ifdef B_HAS_WMS
+#define MAX_PACKET_SIZE 28000
+#else
+#define MAX_PACKET_SIZE 20000
+#endif
+
 ////////// ReorderingPacketBuffer definition //////////
 
 class ReorderingPacketBuffer {
@@ -207,6 +216,11 @@ void MultiFramedRTPSource::doGetNextFram
       fTo += frameSize; fMaxSize -= frameSize;
       fNeedDelivery = True;
     }
+/* PT: BJ/CH team (AY) have applied and need this HACK to get rid of an OOM */
+/* condition, for TCP ASF over IP. Only do this for WMS */
+#ifdef B_HAS_WMS
+    fNeedDelivery = True;
+#endif
   }
 }
 
@@ -328,8 +342,6 @@ void MultiFramedRTPSource::networkReadHa
 
 ////////// BufferedPacket and BufferedPacketFactory implementation /////
 
-#define MAX_PACKET_SIZE 20000
-
 BufferedPacket::BufferedPacket()
   : fPacketSize(MAX_PACKET_SIZE),
     fBuf(new unsigned char[MAX_PACKET_SIZE]),
diff -purN live.2013.12.05/liveMedia/OnDemandServerMediaSubsession.cpp live/liveMedia/OnDemandServerMediaSubsession.cpp
--- live.2013.12.05/liveMedia/OnDemandServerMediaSubsession.cpp	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/OnDemandServerMediaSubsession.cpp	2014-05-19 12:44:05.153037059 +0100
@@ -72,6 +72,32 @@ OnDemandServerMediaSubsession::sdpLines(
   return fSDPLines;
 }
 
+char const*
+OnDemandServerMediaSubsession::sdpLines( const char * filename ) {
+  if (fSDPLines == NULL) {
+    // We need to construct a set of SDP lines that describe this
+    // subsession (as a unicast stream).  To do so, we first create
+    // dummy (unused) source and "RTPSink" objects,
+    // whose parameters we use for the SDP lines:
+    unsigned estBitrate=0; // unused
+    FramedSource* inputSource = createNewStreamSource(0, (unsigned&) filename );
+    if (inputSource == NULL) return NULL; // file not found
+
+    struct in_addr dummyAddr;
+    dummyAddr.s_addr = 0;
+    Groupsock dummyGroupsock(envir(), dummyAddr, 0, 0);
+    unsigned char rtpPayloadType = 96 + trackNumber()-1; // if dynamic
+    RTPSink* dummyRTPSink
+      = createNewRTPSink(&dummyGroupsock, rtpPayloadType, inputSource);
+
+    setSDPLinesFromRTPSink(dummyRTPSink, inputSource, estBitrate);
+    Medium::close(dummyRTPSink);
+    closeStreamSource(inputSource);
+  }
+
+  return fSDPLines;
+}
+
 void OnDemandServerMediaSubsession
 ::getStreamParameters(unsigned clientSessionId,
 		      netAddressBits clientAddress,
@@ -167,11 +193,13 @@ void OnDemandServerMediaSubsession
       increaseSendBufferTo(envir(), rtpGroupsock->socketNum(), rtpBufSize);
     }
 
-    // Set up the state of the stream.  The stream will get started later:
-    streamToken = fLastStreamToken
-      = new StreamState(*this, serverRTPPort, serverRTCPPort, rtpSink, udpSink,
-			streamBitrate, mediaSource,
-			rtpGroupsock, rtcpGroupsock);
+	/* CAD 2014-02-03: call to StreamState() crashes if mediaSource is NULL */
+	if ( mediaSource ) {
+      // Set up the state of the stream.  The stream will get started later:
+      streamToken = fLastStreamToken
+        = new StreamState(*this, serverRTPPort, serverRTCPPort, rtpSink, udpSink,
+          streamBitrate, mediaSource, rtpGroupsock, rtcpGroupsock);
+    }
   }
 
   // Record these destinations as being for this client session id:
diff -purN live.2013.12.05/liveMedia/PassiveServerMediaSubsession.cpp live/liveMedia/PassiveServerMediaSubsession.cpp
--- live.2013.12.05/liveMedia/PassiveServerMediaSubsession.cpp	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/PassiveServerMediaSubsession.cpp	2014-05-19 12:44:17.097337416 +0100
@@ -115,6 +115,11 @@ PassiveServerMediaSubsession::sdpLines()
   return fSDPLines;
 }
 
+char const*
+PassiveServerMediaSubsession::sdpLines( const char *) {
+  return sdpLines();
+}
+
 void PassiveServerMediaSubsession
 ::getStreamParameters(unsigned clientSessionId,
 		      netAddressBits clientAddress,
diff -purN live.2013.12.05/liveMedia/RTSPClient.cpp live/liveMedia/RTSPClient.cpp
--- live.2013.12.05/liveMedia/RTSPClient.cpp	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/RTSPClient.cpp	2014-05-19 13:42:56.294306762 +0100
@@ -298,16 +298,25 @@ RTSPClient::RTSPClient(UsageEnvironment&
 		       int verbosityLevel, char const* applicationName,
 		       portNumBits tunnelOverHTTPPortNum, int socketNumToServer)
   : Medium(env),
-    fVerbosityLevel(verbosityLevel), fCSeq(1), fServerAddress(0),
-    fTunnelOverHTTPPortNum(tunnelOverHTTPPortNum), fUserAgentHeaderStr(NULL), fUserAgentHeaderStrLen(0),
+    fVerbosityLevel(verbosityLevel), fCSeq(1),
+    fTunnelOverHTTPPortNum(tunnelOverHTTPPortNum), fUserAgentHeaderStr(NULL), fUserAgentHeaderStrLen(0), fServerAddress(0), 
     fInputSocketNum(-1), fOutputSocketNum(-1), fBaseURL(NULL), fTCPStreamIdCount(0),
-    fLastSessionId(NULL), fSessionTimeoutParameter(0), fSessionCookieCounter(0), fHTTPTunnelingConnectionIsPending(False) {
+    fLastSessionId(NULL), fSessionTimeoutParameter(0), fSessionCookieCounter(0), fHTTPTunnelingConnectionIsPending(False)
+    // JJ/BRCM 20061026 - save Server type for users of RTSPClient
+    ,fServerTypeStr(NULL)
+    // SS/BRCM 20100210: save RTSP Response Message received for last RTSP Command sent to the server
+    ,fResponseMessageStr(NULL)
+    // SS/BRCM 20100416: pointer to app specified additional HTTP header
+    ,fAdditionHeadersStr(NULL)
+    // CAD/BRCM 20131218 - used to block while waiting for responses for OPTIONS, DESCRIBE, SETUP, PLAY, etc.
+    ,fWatchVariableForSyncInterface(0)
+{
   setBaseURL(rtspURL);
 
   fResponseBuffer = new char[responseBufferSize+1];
   resetResponseBuffer();
 
-  if (socketNumToServer >= 0) {
+  if (socketNumToServer > 0) {
     // This socket number is (assumed to be) already connected to the server.
     // Use it, and arrange to handle responses to requests sent on it:
     fInputSocketNum = fOutputSocketNum = socketNumToServer;
@@ -339,6 +348,13 @@ RTSPClient::~RTSPClient() {
 
   delete[] fResponseBuffer;
   delete[] fUserAgentHeaderStr;
+  // JJ/BRCM 20061026 - save Server type for users of RTSPClient
+  delete fServerTypeStr; fServerTypeStr = NULL;
+  // SS/BRCM 20100210: 
+  delete fResponseMessageStr; fResponseMessageStr = NULL;
+  fAdditionHeadersStr = NULL;
+  // CAD/BRCM 20131218 - used to block while waiting for responses for OPTIONS, DESCRIBE, SETUP, PLAY, etc.
+  fWatchVariableForSyncInterface = 0;
 }
 
 void RTSPClient::reset() {
@@ -351,6 +367,13 @@ void RTSPClient::reset() {
   fCurrentAuthenticator.reset();
 
   delete[] fLastSessionId; fLastSessionId = NULL;
+  // JJ/BRCM 20061026 - save Server type for users of RTSPClient
+  delete fServerTypeStr; fServerTypeStr = NULL;
+  // SS/BRCM 20100210: 
+  delete fResponseMessageStr; fResponseMessageStr = NULL; 
+  fAdditionHeadersStr = NULL;
+  // CAD/BRCM 20131218 - used to block while waiting for responses for OPTIONS, DESCRIBE, SETUP, PLAY, etc.
+  fWatchVariableForSyncInterface = 0;
 }
 
 void RTSPClient::setBaseURL(char const* url) {
@@ -371,7 +394,7 @@ unsigned RTSPClient::sendRequest(Request
     if (!fRequestsAwaitingConnection.isEmpty()) {
       // A connection is currently pending (with at least one enqueued request).  Enqueue this request also:
       connectionIsPending = True;
-    } else if (fInputSocketNum < 0) { // we need to open a connection
+    } else if (fInputSocketNum <= 0) { // we need to open a connection
       int connectResult = openConnection();
       if (connectResult < 0) break; // an error occurred
       else if (connectResult == 0) {
@@ -471,11 +494,12 @@ unsigned RTSPClient::sendRequest(Request
     }
 
     if (send(fOutputSocketNum, cmd, strlen(cmd), 0) < 0) {
-      char const* errFmt = "%s send() failed: ";
+      char const* errFmt = "RTSPClient::sendRequest: %s send() failed: ";
       unsigned const errLength = strlen(errFmt) + strlen(request->commandName());
       char* err = new char[errLength];
       sprintf(err, errFmt, request->commandName());
       envir().setResultErrMsg(err);
+      fprintf(stderr,"%s: ERROR send(socket %d); cmd (%s)\n", __FUNCTION__, fOutputSocketNum, cmd );
       delete[] err;
       break;
     }
@@ -1504,6 +1528,16 @@ void RTSPClient::handleResponseBytes(int
     char* bodyStart = NULL;
     unsigned numBodyBytes = 0;
     responseSuccess = False;
+
+  /* PT: BJ/CH team (AY) have applied and need this HACK to get rid of   */
+  /* a macroblocking condition, for TCP ASF over IP only do this for WMS */
+
+#ifdef B_HAS_WMS
+  if(!strcasecmp(tag, "GET_PARAMETER")) {
+      return True;
+  }
+#endif
+  
     do {
       headerDataCopy = new char[responseBufferSize];
       strncpy(headerDataCopy, fResponseBuffer, fResponseBytesAlreadySeen);
@@ -1579,6 +1613,9 @@ void RTSPClient::handleResponseBytes(int
 	  // Note: we accept "Allow:" instead of "Public:", so that "OPTIONS" requests made to HTTP servers will work.
 	} else if (checkForHeader(lineStart, "Location:", 9, headerParamsStr)) {
 	  setBaseURL(headerParamsStr);
+      // JJ/BRCM 20061026 - save Server type for users of RTSPClient
+	} else if (checkForHeader(lineStart, "Server: ", 8, headerParamsStr)) {
+        delete fServerTypeStr; fServerTypeStr = strDup(&lineStart[8]);
 	}
       }
       if (!reachedEndOfHeaders) break; // an error occurred
@@ -1633,6 +1670,9 @@ void RTSPClient::handleResponseBytes(int
 	if (responseCode == 200) {
 	  // Do special-case response handling for some commands:
 	  if (strcmp(foundRequest->commandName(), "SETUP") == 0) {
+          // SS/BRCM 20100210: 
+          delete[] fResponseMessageStr; fResponseMessageStr = strDup(nextLineStart);
+
 	    if (!handleSETUPResponse(*foundRequest->subsession(), sessionParamsStr, transportParamsStr, foundRequest->booleanFlags()&0x1)) break;
 	  } else if (strcmp(foundRequest->commandName(), "PLAY") == 0) {
 	    if (!handlePLAYResponse(*foundRequest->session(), *foundRequest->subsession(), scaleParamsStr, rangeParamsStr, rtpInfoParamsStr)) break;
@@ -1640,7 +1680,10 @@ void RTSPClient::handleResponseBytes(int
 	    if (!handleTEARDOWNResponse(*foundRequest->session(), *foundRequest->subsession())) break;
 	  } else if (strcmp(foundRequest->commandName(), "GET_PARAMETER") == 0) {
 	    if (!handleGET_PARAMETERResponse(foundRequest->contentStr(), bodyStart)) break;
-	  }
+      } else if (strcmp(foundRequest->commandName(), "OPTONS") == 0) {
+          // SS/BRCM 20100210: 
+          delete[] fResponseMessageStr; fResponseMessageStr = strDup(nextLineStart);
+      }
 	} else if (responseCode == 401 && handleAuthenticationFailure(wwwAuthenticateParamsStr)) {
 	  // We need to resend the command, with an "Authorization:" header:
 	  needToResendCommand = True;
diff -purN live.2013.12.05/liveMedia/RTSPServer.cpp live/liveMedia/RTSPServer.cpp
--- live.2013.12.05/liveMedia/RTSPServer.cpp	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/RTSPServer.cpp	2014-05-19 12:47:22.433999051 +0100
@@ -20,11 +20,26 @@ along with this library; if not, write t
 
 #include "RTSPServer.hh"
 #include "RTSPCommon.hh"
-#include "RTSPRegisterSender.hh"
-#include "ProxyServerMediaSession.hh"
 #include "Base64.hh"
-#include <GroupsockHelper.hh>
+#include "liveMedia.hh"
+#include "BasicUsageEnvironment.hh"
+#include "GroupsockHelper.hh"
+#ifdef DEBUG
+#include <pthread.h>
+#endif /* DEBUG */
+#include <time.h> // for "strftime()" and "gmtime()"
+#include <net/if.h>
+
+ServerMediaSession * AddRtspStream ( char const* streamName, char const* inputFileName, char const* descriptionString, RTSPServer* rtspServer );
+int SocketOptions ( int fd );
+
+char * formatRemoteFilename ( unsigned long int Ctx )
+{
+    static char filename[32];
+    sprintf(filename, "brcmsat.%lx", Ctx );
 
+    return filename;
+}
 ////////// RTSPServer implementation //////////
 
 RTSPServer*
@@ -1362,7 +1377,8 @@ RTSPServer::RTSPClientSession
 
 RTSPServer::RTSPClientSession::~RTSPClientSession() {
   // Turn off any liveness checking:
-  envir().taskScheduler().unscheduleDelayedTask(fLivenessCheckTask);
+	/* CAD 2014-04-08: this function crashes when you ctrl-c BIP_RtspServer derived class */
+  if(fLivenessCheckTask) envir().taskScheduler().unscheduleDelayedTask(fLivenessCheckTask);
   
   // Remove ourself from the server's 'client sessions' hash table before we go:
   char sessionIdStr[9];
@@ -2246,3 +2262,8 @@ UserAuthenticationDatabase* RTSPServerWi
 
   return RTSPServer::getAuthenticationDatabaseForCommand(cmdName);
 }
+
+const char * LiveMediaGetVersion ( void )
+{
+    return LIVEMEDIA_LIBRARY_VERSION_STRING;
+}
diff -purN live.2013.12.05/liveMedia/ServerMediaSession.cpp live/liveMedia/ServerMediaSession.cpp
--- live.2013.12.05/liveMedia/ServerMediaSession.cpp	2013-12-05 22:35:43.000000000 +0000
+++ live/liveMedia/ServerMediaSession.cpp	2014-05-19 12:48:02.483006176 +0100
@@ -89,6 +89,10 @@ ServerMediaSession::~ServerMediaSession(
 
 Boolean
 ServerMediaSession::addSubsession(ServerMediaSubsession* subsession) {
+    if (subsession == NULL) {
+      return false;
+    }
+
   if (subsession->fParentSession != NULL) return False; // it's already used
 
   if (fSubsessionsTail == NULL) {
@@ -440,3 +444,9 @@ ServerMediaSubsession::rangeSDPLine() co
     return strDup(buf);
   }
 }
+
+char const*
+ServerMediaSubsession::infoSDPString() const {
+  if (fParentSession == NULL) return NULL;
+  return fParentSession->sInfoSDPString();
+}
diff -purN live.2013.12.05/Makefile live/Makefile
--- live.2013.12.05/Makefile	2014-05-19 13:52:09.120341556 +0100
+++ live/Makefile	2014-05-19 13:35:02.734305285 +0100
@@ -36,13 +36,12 @@ B_REFSW_CFLAGS += -D_FILE_OFFSET_BITS=64
 DESTDIR = ../../../../../../nexus/bin
 # LIBDIR is used if you want the libraries to be in a subdirectory ... like /usr/local/lib
 LIBDIR = 
-PREFIX = 
 
 B_REFSW_ARCH ?= mipsel-uclibc
-ifeq ($(filter ${B_REFSW_ARCH}, mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
-CROSS_COMPILE ?= mips-linux-uclibc-
+ifeq ($(filter ${B_REFSW_ARCH}, arm-linux mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
+CROSS_COMPILE ?= $(B_REFSW_ARCH)-
 else
-CROSS_COMPILE ?= mipsel-linux-uclibc-
+CROSS_COMPILE ?= mipsel-linux-
 endif
 
 COMPILE_OPTS =		$(INCLUDES) -I. -DSOCKLEN_T=socklen_t -DNO_STRSTREAM=1 -DNO_SSTREAM=1 $(B_REFSW_CFLAGS) -O2 -DXLOCALE_NOT_USED=1 -fPIC
diff -purN live.2013.12.05/mediaServer/Makefile live/mediaServer/Makefile
--- live.2013.12.05/mediaServer/Makefile	2014-05-19 13:52:09.112341351 +0100
+++ live/mediaServer/Makefile	2014-05-19 13:31:52.581483383 +0100
@@ -42,13 +42,12 @@ B_REFSW_CFLAGS += -D_FILE_OFFSET_BITS=64
 DESTDIR = ../../../../../../nexus/bin
 # LIBDIR is used if you want the libraries to be in a subdirectory ... like /usr/local/lib
 LIBDIR = 
-PREFIX = 
 
 B_REFSW_ARCH ?= mipsel-uclibc
-ifeq ($(filter ${B_REFSW_ARCH}, mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
-CROSS_COMPILE ?= mips-linux-uclibc-
+ifeq ($(filter ${B_REFSW_ARCH}, arm-linux mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
+CROSS_COMPILE ?= $(B_REFSW_ARCH)-
 else
-CROSS_COMPILE ?= mipsel-linux-uclibc-
+CROSS_COMPILE ?= mipsel-linux-
 endif
 
 COMPILE_OPTS =		$(INCLUDES) -I. -DSOCKLEN_T=socklen_t -DNO_STRSTREAM=1 -DNO_SSTREAM=1 $(B_REFSW_CFLAGS) -O2 -DXLOCALE_NOT_USED=1 -fPIC
@@ -79,7 +78,7 @@ INSTALL2 =		install_shared_libraries
 
 MEDIA_SERVER = live555MediaServer$(EXE)
 
-PREFIX = /usr/local
+
 ALL = $(MEDIA_SERVER)
 all: $(ALL)
 
diff -purN live.2013.12.05/proxyServer/Makefile live/proxyServer/Makefile
--- live.2013.12.05/proxyServer/Makefile	2014-05-19 13:52:09.116341455 +0100
+++ live/proxyServer/Makefile	2014-05-19 13:34:24.153326896 +0100
@@ -42,13 +42,12 @@ B_REFSW_CFLAGS += -D_FILE_OFFSET_BITS=64
 DESTDIR = ../../../../../../nexus/bin
 # LIBDIR is used if you want the libraries to be in a subdirectory ... like /usr/local/lib
 LIBDIR = 
-PREFIX = 
 
 B_REFSW_ARCH ?= mipsel-uclibc
-ifeq ($(filter ${B_REFSW_ARCH}, mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
-CROSS_COMPILE ?= mips-linux-uclibc-
+ifeq ($(filter ${B_REFSW_ARCH}, arm-linux mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
+CROSS_COMPILE ?= $(B_REFSW_ARCH)-
 else
-CROSS_COMPILE ?= mipsel-linux-uclibc-
+CROSS_COMPILE ?= mipsel-linux-
 endif
 
 COMPILE_OPTS =		$(INCLUDES) -I. -DSOCKLEN_T=socklen_t -DNO_STRSTREAM=1 -DNO_SSTREAM=1 $(B_REFSW_CFLAGS) -O2 -DXLOCALE_NOT_USED=1 -fPIC
@@ -79,7 +78,7 @@ INSTALL2 =		install_shared_libraries
 
 PROXY_SERVER = live555ProxyServer$(EXE)
 
-PREFIX = /usr/local
+
 ALL = $(PROXY_SERVER)
 all: $(ALL)
 
diff -purN live.2013.12.05/testProgs/Makefile live/testProgs/Makefile
--- live.2013.12.05/testProgs/Makefile	2014-05-19 13:52:09.108341245 +0100
+++ live/testProgs/Makefile	2014-05-19 13:34:43.377814377 +0100
@@ -42,13 +42,12 @@ B_REFSW_CFLAGS += -D_FILE_OFFSET_BITS=64
 DESTDIR = ../../../../../../nexus/bin
 # LIBDIR is used if you want the libraries to be in a subdirectory ... like /usr/local/lib
 LIBDIR = 
-PREFIX = 
 
 B_REFSW_ARCH ?= mipsel-uclibc
-ifeq ($(filter ${B_REFSW_ARCH}, mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
-CROSS_COMPILE ?= mips-linux-uclibc-
+ifeq ($(filter ${B_REFSW_ARCH}, arm-linux mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
+CROSS_COMPILE ?= $(B_REFSW_ARCH)-
 else
-CROSS_COMPILE ?= mipsel-linux-uclibc-
+CROSS_COMPILE ?= mipsel-linux-
 endif
 
 COMPILE_OPTS =		$(INCLUDES) -I. -DSOCKLEN_T=socklen_t -DNO_STRSTREAM=1 -DNO_SSTREAM=1 $(B_REFSW_CFLAGS) -O2 -DXLOCALE_NOT_USED=1 -fPIC
@@ -88,7 +87,7 @@ UNICAST_APPS = $(UNICAST_STREAMER_APPS)
 
 MISC_APPS = testMPEG1or2Splitter$(EXE) testMPEG1or2ProgramToTransportStream$(EXE) testH264VideoToTransportStream$(EXE) MPEG2TransportStreamIndexer$(EXE) testMPEG2TransportStreamTrickPlay$(EXE) registerRTSPStream$(EXE)
 
-PREFIX = /usr/local
+
 ALL = $(MULTICAST_APPS) $(UNICAST_APPS) $(MISC_APPS)
 all: $(ALL)
 
diff -purN live.2013.12.05/UsageEnvironment/Makefile live/UsageEnvironment/Makefile
--- live.2013.12.05/UsageEnvironment/Makefile	2014-05-19 13:52:09.096340990 +0100
+++ live/UsageEnvironment/Makefile	2014-05-19 13:29:23.901712248 +0100
@@ -1,6 +1,6 @@
 INCLUDES = -Iinclude -I../groupsock/include
-PREFIX = /usr/local
-LIBDIR = $(PREFIX)/lib
+
+
 ##### Change the following for your environment:
 # 'CURRENT':'REVISION':'AGE' are updated - whenever a library changes - as follows:
 # The library code changes, but without any changes to the API (i.e., interfaces) => increment REVISION
@@ -39,13 +39,12 @@ B_REFSW_CFLAGS += -D_FILE_OFFSET_BITS=64
 DESTDIR = ../../../../../../nexus/bin
 # LIBDIR is used if you want the libraries to be in a subdirectory ... like /usr/local/lib
 LIBDIR = 
-PREFIX = 
 
 B_REFSW_ARCH ?= mipsel-uclibc
-ifeq ($(filter ${B_REFSW_ARCH}, mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
-CROSS_COMPILE ?= mips-linux-uclibc-
+ifeq ($(filter ${B_REFSW_ARCH}, arm-linux mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
+CROSS_COMPILE ?= $(B_REFSW_ARCH)-
 else
-CROSS_COMPILE ?= mipsel-linux-uclibc-
+CROSS_COMPILE ?= mipsel-linux-
 endif
 
 COMPILE_OPTS =		$(INCLUDES) -I. -DSOCKLEN_T=socklen_t -DNO_STRSTREAM=1 -DNO_SSTREAM=1 $(B_REFSW_CFLAGS) -O2 -DXLOCALE_NOT_USED=1 -fPIC
@@ -58,20 +57,15 @@ CPLUSPLUS_COMPILER =	$(B_REFSW_CROSS_COM
 CPLUSPLUS_FLAGS =	$(COMPILE_OPTS) -Wall -DBSD=1 -fexceptions $(CPPFLAGS) $(CXXFLAGS)
 OBJ =			o
 LINK =			$(B_REFSW_CROSS_COMPILE)c++ -o
-LINK_OPTS =		-L. $(LDFLAGS) $(ANDROID_LDFLAGS)
+LINK_OPTS =		-L.
 CONSOLE_LINK_OPTS =	$(LINK_OPTS)
-LIBRARY_LINK =		$(B_REFSW_CROSS_COMPILE)gcc -o 
-SHORT_LIB_SUFFIX =	so.$(shell expr $($(NAME)_VERSION_CURRENT) - $($(NAME)_VERSION_AGE))
-LIB_SUFFIX =	 	$(SHORT_LIB_SUFFIX).$($(NAME)_VERSION_AGE).$($(NAME)_VERSION_REVISION)
-LIBRARY_LINK_OPTS =	-shared $(LDFLAGS) -nostdlib -Wl,--no-undefined,--gc-sections -L$(ANDROID_LIB) \
-                    $(ANDROID_LIB)/crtbegin_so.o $(ANDROID_LIB)/crtend_so.o \
-                    -Wl,--no-whole-archive -Wl,-dy,-ldl,-lstdc++,-lcutils,-lc,-lm \
-                    -Wl,--whole-archive -Wl,-lgcc,-lsupc++
-
+LIBRARY_LINK =		$(B_REFSW_CROSS_COMPILE)ld -o
+LIBRARY_LINK_OPTS =	$(LINK_OPTS) -r -Bstatic
+LIB_SUFFIX =			a
 LIBS_FOR_CONSOLE_APPLICATION =
 LIBS_FOR_GUI_APPLICATION =
 EXE =
-INSTALL2 =		install_shared_libraries
+
 ##### End of variables to change
 
 NAME = libUsageEnvironment
@@ -105,7 +99,7 @@ install1: $(USAGE_ENVIRONMENT_LIB)
 	  install -m 644 include/*.hh $(DESTDIR)$(PREFIX)/include/UsageEnvironment
 	  install -m 644 $(USAGE_ENVIRONMENT_LIB) $(DESTDIR)$(LIBDIR)
 install_shared_libraries: $(USAGE_ENVIRONMENT_LIB)
-	  ln -s $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).$(SHORT_LIB_SUFFIX)
-	  ln -s $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).so
+	  ln -fs $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).$(SHORT_LIB_SUFFIX)
+	  ln -fs $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).so
 
 ##### Any additional, platform-specific rules come here:
