diff -Naur wpa_supplicant-2.5-orig/wpa_supplicant/Makefile wpa_supplicant-2.5/wpa_supplicant/Makefile
--- wpa_supplicant-2.5-orig/wpa_supplicant/Makefile	2016-08-16 11:13:06.900293858 -0700
+++ wpa_supplicant-2.5/wpa_supplicant/Makefile	2016-08-16 11:07:40.762408521 -0700
@@ -1691,9 +1691,11 @@
 
 LIBCTRL += ../src/common/wpa_ctrl.o
 LIBCTRL += ../src/utils/os_$(CONFIG_OS).o
+LIBCTRL += ../src/utils/common.o
 LIBCTRL += ../src/utils/wpa_debug.o
 LIBCTRLSO += ../src/common/wpa_ctrl.c
 LIBCTRLSO += ../src/utils/os_$(CONFIG_OS).c
+LIBCTRLSO += ../src/utils/common.c
 LIBCTRLSO += ../src/utils/wpa_debug.c
 
 libwpa_client.a: $(LIBCTRL)
