diff -ru SSPK-release-V1_5.orig/Source/Components/dirs SSPK-release-V1_5/Source/Components/dirs
--- SSPK-release-V1_5.orig/Source/Components/dirs	2012-10-11 14:19:10.841178000 -0400
+++ SSPK-release-V1_5/Source/Components/dirs	2012-10-11 14:22:17.472221000 -0400
@@ -9,3 +9,4 @@
   Streamer \
   Transport \
   Utilities \
+
diff -ru SSPK-release-V1_5.orig/Source/Platform/Common/dirs SSPK-release-V1_5/Source/Platform/Common/dirs
--- SSPK-release-V1_5.orig/Source/Platform/Common/dirs	2012-10-11 14:19:12.226382000 -0400
+++ SSPK-release-V1_5/Source/Platform/Common/dirs	2012-10-11 14:22:17.473224000 -0400
@@ -1 +1,2 @@
 dirs = \
+
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/Build/MspkBase.mk SSPK-release-V1_5/Source/Platform/LinuxSample/Build/MspkBase.mk
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/Build/MspkBase.mk	2012-10-11 14:19:12.316381000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/Build/MspkBase.mk	2012-10-11 10:43:55.744622000 -0400
@@ -96,19 +96,6 @@
 endif
 
 #
-# Ignore default include and lib files if we're
-# using a supplied toolchain
-#
-ifneq '$(MSPKOPTION_USE_SYSTEM_NATIVE_TOOLCHAIN)' '1'
-CFLAGS += -nostdinc
-CPPFLAGS += -nostdinc++
-CXXFLAGS += -nostdinc++
-CXXFLAGS += -fno-rtti
-PLATLINKFLAGS += -nostdlib -nostartfiles
-endif
-
-
-#
 # If we want listing files, also make the linker give verbose output
 # (will only go to the log file, so it won't pollute the default build
 # output)
@@ -118,118 +105,11 @@
 endif
 
 
-#
-# If we have not specified a target distribution to build for, use
-# the default
-#
-ifeq '$(MSPK_LINUXSAMPLE_TARGET_DISTRO)' ''
-MSPK_LINUXSAMPLE_TARGET_DISTRO=Fedora10
-endif
-
-
-#
-# Location of the toolchain, and some information about the toolchain
-#
-ifneq '$(MSPKOPTION_USE_SYSTEM_NATIVE_TOOLCHAIN)' '1'
-TOOLCHAIN=$(MSPK_TOOLCHAIN)/Toolchain/$(MSPK_PLATFORM)/$(MSPK_LINUXSAMPLE_TARGET_DISTRO)
-TOOLCHAIN_TARGET=i386-redhat-linux
-
-ifeq '$(MSPK_LINUXSAMPLE_TARGET_DISTRO)' 'Fedora10'
-TOOLCHAIN_VERSION=4.3.2
-endif
-
-ifeq '$(MSPK_LINUXSAMPLE_TARGET_DISTRO)' 'Fedora6'
-TOOLCHAIN_VERSION=4.1.1
-endif
-
-ifeq '$(MSPK_LINUXSAMPLE_TARGET_DISTRO)' 'Ubuntu8.04'
-TOOLCHAIN_VERSION=4.2.4
-endif
-
-else
-#
-# If we are using the native toolchain, then it should come off the root
-#
 TOOLCHAIN=
-endif
-
-
-#
-# Add the toolchain's include file path
-#
-ifneq '$(MSPKOPTION_USE_SYSTEM_NATIVE_TOOLCHAIN)' '1'
-C_SYSTEM_INCLUDES += $(TOOLCHAIN)/usr/lib/gcc/$(TOOLCHAIN_TARGET)/$(TOOLCHAIN_VERSION)/include
-C_SYSTEM_INCLUDES += $(TOOLCHAIN)/usr/local/include
-C_SYSTEM_INCLUDES += $(TOOLCHAIN)/usr/include
-C_SYSTEM_INCLUDES += $(TOOLCHAIN)/usr/include/linux
-endif
-
-
-#
-# Include folders needed for GStreamer
-#
-INCLUDES += $(TOOLCHAIN)/usr/include/gstreamer-0.10;
-INCLUDES += $(TOOLCHAIN)/usr/include/glib-2.0;
-INCLUDES += $(TOOLCHAIN)/usr/include/libxml2;
-INCLUDES += $(TOOLCHAIN)/usr/lib/i386-linux-gnu/glib-2.0/include;
-INCLUDES += $(TOOLCHAIN)/usr/lib/glib-2.0/include;
-
-
-#
-# Linker options to set the system folders and system libraries for
-# the toolchain
-#
-ifneq '$(MSPKOPTION_USE_SYSTEM_NATIVE_TOOLCHAIN)' '1'
-
-
-#
-# Library folders to look in
-#
-PLATLINKFLAGS += -L$(TOOLCHAIN)/usr/i386-redhat-linux/lib
-PLATLINKFLAGS += -L$(TOOLCHAIN)/usr/local/lib
-PLATLINKFLAGS += -L$(TOOLCHAIN)/lib
-PLATLINKFLAGS += -L$(TOOLCHAIN)/usr/lib
-
-
-#
-# Shared object library folders to look in
-#
-PLATLINKFLAGS += -Wl,-rpath,$(TOOLCHAIN)/lib
-PLATLINKFLAGS += -Wl,-rpath,$(TOOLCHAIN)/usr/lib
-
-
-#
-# CRT startup modules to link
-#
-CRT_STARTUP_OBJS += $(TOOLCHAIN)/usr/lib/gcc/$(TOOLCHAIN_TARGET)/$(TOOLCHAIN_VERSION)/../../../crt1.o
-CRT_STARTUP_OBJS += $(TOOLCHAIN)/usr/lib/gcc/$(TOOLCHAIN_TARGET)/$(TOOLCHAIN_VERSION)/../../../crti.o
-CRT_STARTUP_OBJS += $(TOOLCHAIN)/usr/lib/gcc/$(TOOLCHAIN_TARGET)/$(TOOLCHAIN_VERSION)/../../../crtn.o
-CRT_STARTUP_OBJS += $(TOOLCHAIN)/usr/lib/gcc/$(TOOLCHAIN_TARGET)/$(TOOLCHAIN_VERSION)/crtbegin.o
-CRT_STARTUP_OBJS += $(TOOLCHAIN)/usr/lib/gcc/$(TOOLCHAIN_TARGET)/$(TOOLCHAIN_VERSION)/crtend.o
-
-
 #
 # Libraries to link
 #
-PLATLINKFLAGS += -lgcc
-PLATLINKFLAGS += -lc
-PLATLINKFLAGS += -lc_nonshared
-PLATLINKFLAGS += -lpthread
-PLATLINKFLAGS += -lpthread_nonshared
-PLATLINKFLAGS += -lrt
-PLATLINKFLAGS += -lgcc_s
-PLATLINKFLAGS += -lgstaudio-0.10
-PLATLINKFLAGS += -lgstbase-0.10
-PLATLINKFLAGS += -lgstinterfaces-0.10
-PLATLINKFLAGS += -lgstreamer-0.10
-PLATLINKFLAGS += -lX11
-
-
-#$(warning LD_LIBRARY_PATH: $(LD_LIBRARY_PATH))
-LD_LIBRARY_PATH = :$(TOOLCHAIN)/usr/lib:$(TOOLCHAIN)/usr/lib:$(TOOLCHAIN)/lib
-#$(warning LD_LIBRARY_PATH: $(LD_LIBRARY_PATH))
-
-endif
+PLATLINKFLAGS += -lpthread ${LDFLAGS_EXTRA}
 
 
 #
@@ -492,6 +372,14 @@
 endif
 endif
 
+ifdef IN_HAL_IMPL
+CFLAGS += ${HAL_CFLAGS}
+endif
+
+ifdef IN_DAL_IMPL
+CFLAGS += ${DAL_CFLAGS}
+endif
+
 #
 # Parse through the semicolon-delimited $INCLUDES
 # variable, and build a new variable with each item
@@ -537,12 +425,7 @@
 #
 # Malloc Record
 #
-C_DEFINES += -DMALLOC_RECORD
-
-#
-# AVRenderer
-#
-C_DEFINES += -DSCALE_PTS_TO_1KHZ
+# C_DEFINES += -DMALLOC_RECORD
 
 #
 # Convert Tests' requests for certain PALs to #defines
@@ -566,6 +449,7 @@
 #
 # C++ definitions should be based on C definitions
 #
+
 CPPFLAGS += $(CFLAGS) $(CPP_SYSTEM_INCLUDE_FLAGS)
 CXXFLAGS += $(CFLAGS) $(CPP_SYSTEM_INCLUDE_FLAGS)
 
@@ -575,6 +459,8 @@
 CPPFLAGS += -std=c++0x
 CXXFLAGS += -std=c++0x
 
+
+
 MSPK_PAL_LIBS = \
 	$(MSPK_OBJ_ROOT)/Platform/$(MSPK_PLATFORM)/HALimpl/Crypto/Crypto.a \
 	$(MSPK_OBJ_ROOT)/Platform/$(MSPK_PLATFORM)/HALimpl/Decoder/Decoder.a \
@@ -593,52 +479,7 @@
 	$(MSPK_PLATFORM_PATH)/TestFramework/main \
 
 
-#
-# Starting with Ubuntu 11.04 some libs are in the /usr/lib/i386-linux-gnu sub-directory.
-# Starting with Ubuntu 12.04 all of the libs are now moved down.
-#
-ifeq '$(wildcard /usr/lib/i386-linux-gnu/*.so)' ''
-# Ubuntu 10.04 LTS:
-MSPK_PAL_TESTFRAMEWORK_DYNAMIC_LIBS = \
-	$(TOOLCHAIN)/usr/lib/librt.so \
-	$(TOOLCHAIN)/lib/libpthread.so.0 \
-	$(TOOLCHAIN)/usr/lib/libstdc++.so.6 \
-	$(TOOLCHAIN)/usr/lib/libgstreamer-0.10.so \
-	$(TOOLCHAIN)/usr/lib/libgstinterfaces-0.10.so \
-	$(TOOLCHAIN)/usr/lib/libgstbase-0.10.so \
-	$(TOOLCHAIN)/usr/lib/libgstaudio-0.10.so \
-	$(TOOLCHAIN)/usr/lib/libX11.so \
-	$(TOOLCHAIN)/usr/lib/libz.so \
-
-else
-ifeq '$(wildcard /usr/lib/i386-linux-gnu/libgst*.so)' ''
-# Ubuntu 11.04/11.10:
-MSPK_PAL_TESTFRAMEWORK_DYNAMIC_LIBS = \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/librt.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libpthread.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libstdc++.so.6 \
-	$(TOOLCHAIN)/usr/lib/libgstreamer-0.10.so \
-	$(TOOLCHAIN)/usr/lib/libgstinterfaces-0.10.so \
-	$(TOOLCHAIN)/usr/lib/libgstbase-0.10.so \
-	$(TOOLCHAIN)/usr/lib/libgstaudio-0.10.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libX11.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libz.so \
-
-else
-# Ubuntu 12.04 LTS:
-MSPK_PAL_TESTFRAMEWORK_DYNAMIC_LIBS = \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/librt.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libpthread.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libstdc++.so.6 \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libgstreamer-0.10.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libgstinterfaces-0.10.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libgstbase-0.10.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libgstaudio-0.10.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libX11.so \
-	$(TOOLCHAIN)/usr/lib/i386-linux-gnu/libz.so \
-
-endif
-endif
+MSPK_PAL_TESTFRAMEWORK_DYNAMIC_LIBS = -lpthread  -lstdc++
 
 #
 # the default build should build the target
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/Build/MspkMakeApp.mk SSPK-release-V1_5/Source/Platform/LinuxSample/Build/MspkMakeApp.mk
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/Build/MspkMakeApp.mk	2012-10-11 14:19:12.320384000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/Build/MspkMakeApp.mk	2012-10-11 14:22:17.477221000 -0400
@@ -24,6 +24,8 @@
 #
 TARGETNAME = $(MSPK_TARGETNAME)
 
+CFLAGS += ${HAL_CFLAGS}
+
 
 #
 # Convert back-slashes (Windows-style) in paths to
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/Compatibility/StringSafe/StringSafe.c SSPK-release-V1_5/Source/Platform/LinuxSample/Compatibility/StringSafe/StringSafe.c
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/Compatibility/StringSafe/StringSafe.c	2012-10-11 14:19:12.364381000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/Compatibility/StringSafe/StringSafe.c	2012-10-11 14:22:17.479221000 -0400
@@ -475,6 +475,7 @@
     // leave the last space for the null terminator
     cchMax = cchDest - 1;
 
+    snprintf((char *)pszDest, cchDest, "", fmt, argList);
     iRet = vswprintf(pszDest, cchDest, fmt, argList);
     if(iRet >= (int)cchDest)
     {
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/Compatibility/dirs SSPK-release-V1_5/Source/Platform/LinuxSample/Compatibility/dirs
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/Compatibility/dirs	2012-10-11 14:19:12.335382000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/Compatibility/dirs	2012-10-11 14:22:17.480224000 -0400
@@ -1,3 +1,4 @@
 dirs = \
   StringSafe \
   WinEmulate \
+
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/DRMimpl/XDrmPR/build.pk SSPK-release-V1_5/Source/Platform/LinuxSample/DRMimpl/XDrmPR/build.pk
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/DRMimpl/XDrmPR/build.pk	2012-10-11 14:19:12.432384000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/DRMimpl/XDrmPR/build.pk	2012-10-11 10:09:41.367208000 -0400
@@ -4,16 +4,13 @@
 #
 #*****************************************************************
 
-# Note: the MSPK_TARGETNAME is *not* XDrmPR because this alternate 
-#       build is substituting for the stub library build from XDrm
 MSPK_TARGETNAME=XDrm
-
 MSPK_TARGETPATH=Platform/$(MSPK_PLATFORM)/DRMimpl/$(MSPK_TARGETNAME)
 
+IN_DAL_IMPL = 1
 
 MSPK_CDEFINES = \
 	-DMSPK_USING_STD_TEMPLATE_LIBS \
-	-DDRM_BUILD_PROFILE=DRM_BUILD_PROFILE_OEM \
 
 
 MSPK_DBG_CDEFINES= \
@@ -26,20 +23,6 @@
 
 MSPK_INCLUDES= \
 	$(MSPK_PAL_INCLUDES); \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/inc; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/results; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/oem/common/inc; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/crypto/bignum; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/blackbox; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/keyfile; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/keyfileformat; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/xmr; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/xbinary; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/oem/ansi/inc; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/certs; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/certs/nd; \
-	$(MSPK_ROOT)/Platform/LinuxSample/DRMimpl/WMDRM-MD/ndtinc; \
-
 
 # not IN_PAL_IMPL because IXDrm is C++
 
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/DRMimpl/dirs SSPK-release-V1_5/Source/Platform/LinuxSample/DRMimpl/dirs
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/DRMimpl/dirs	2012-10-11 14:19:12.392384000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/DRMimpl/dirs	2012-09-24 11:51:53.905321000 -0400
@@ -1,2 +1,11 @@
+
+ifeq ($(MSDRM_PRDY_SUPPORT), y)
+drm_folder := XDrmPR
+else
+drm_folder := XDrm
+endif
+
 dirs = \
-  XDrm \
+  $(drm_folder) \
+
+
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/HALimpl/Decoder/build.pk SSPK-release-V1_5/Source/Platform/LinuxSample/HALimpl/Decoder/build.pk
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/HALimpl/Decoder/build.pk	2012-10-11 14:19:12.485382000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/HALimpl/Decoder/build.pk	2012-10-11 14:22:17.482222000 -0400
@@ -7,6 +7,7 @@
 MSPK_TARGETNAME=Decoder
 MSPK_TARGETPATH=Platform/$(MSPK_PLATFORM)/HALimpl/$(MSPK_TARGETNAME)
 
+IN_HAL_IMPL = 1
 
 MSPK_CDEFINES= \
 
@@ -15,8 +16,7 @@
 
 
 MSPK_SOURCES= \
-	CAVRendererGst.cpp \
-	IPTVDecoderHal.cpp \
+	IPTVDecoderHal.cpp
 
 
 MSPK_INCLUDES= \
@@ -24,5 +24,4 @@
 	$(MSPK_PLATFORM_PATH)/PALimpl/include; \
 	
 
-# not IN_PAL_IMPL because HALs are C++
 
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/HALimpl/PhysMemMgr/build.pk SSPK-release-V1_5/Source/Platform/LinuxSample/HALimpl/PhysMemMgr/build.pk
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/HALimpl/PhysMemMgr/build.pk	2012-10-11 14:19:12.528383000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/HALimpl/PhysMemMgr/build.pk	2012-10-11 14:25:05.681173000 -0400
@@ -7,6 +7,7 @@
 MSPK_TARGETNAME=PhysMemMgr
 MSPK_TARGETPATH=Platform/$(MSPK_PLATFORM)/HALimpl/$(MSPK_TARGETNAME)
 
+IN_HAL_IMPL = 1
 
 MSPK_CDEFINES= \
 
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/PALimpl/Sockets/pkSockets.c SSPK-release-V1_5/Source/Platform/LinuxSample/PALimpl/Sockets/pkSockets.c
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/PALimpl/Sockets/pkSockets.c	2012-10-11 14:19:12.684262000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/PALimpl/Sockets/pkSockets.c	2012-10-11 14:22:17.484221000 -0400
@@ -37,6 +37,9 @@
 
 
 
+#include <stdio.h>
+#define XXX_TRACE(x)
+
 static void s_SetThreadLastSocketError(int32_t iLastError)
 {
     pkRESULT pkRet = pkS_OK;
@@ -237,6 +240,7 @@
     {
         s_SetThreadLastSocketError(SOCKET_SUCCESS);
     }
+    XXX_TRACE(("\nRecv %d\n", retval));
     return retval;
 }
 
@@ -291,6 +295,7 @@
         SetThreadLastSocketError_priv(pkE_INVALIDARG);
         return retval;
     }
+    XXX_TRACE(("\nRecv Multi%d\n", retval));
 
     retval = send((int)hSocket, buf, len, (int)flags|MSG_NOSIGNAL);
 
@@ -896,7 +901,13 @@
     )
 {
     //we may want to do more than this, there is concern that this call my block forever
-    return getaddrinfo(hostname, servname, (struct addrinfo *) hints, (struct addrinfo **) ai);
+    //
+    /* struct addrinfo * h = (struct addrinfo *)hints; */
+    int rc;
+    hints = NULL; /* EGLIBS doesn't like hints */
+    rc = getaddrinfo(hostname, servname, (struct addrinfo *) hints, (struct addrinfo **) ai);
+    /* printf("getaddrinfo: %s %s %#x %#x %#x %#x %#x %s %#x %#x -> %d\n", hostname, servname, h->ai_flags, h->ai_family, h->ai_socktype, h->ai_protocol, h->ai_addrlen, h->ai_canonname, h->ai_addr, h->ai_next, rc); */
+    return rc;
 }
 
 void pkAPI Socket_FreeAddrInfo(struct SOCKET_ADDR_INFO *ai)
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/TestFramework/dirs SSPK-release-V1_5/Source/Platform/LinuxSample/TestFramework/dirs
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/TestFramework/dirs	2012-10-11 14:19:12.689260000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/TestFramework/dirs	2012-10-11 14:22:17.486221000 -0400
@@ -1,2 +1,3 @@
 dirs = \
   lib \
+
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/TestFramework/lib/main.c SSPK-release-V1_5/Source/Platform/LinuxSample/TestFramework/lib/main.c
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/TestFramework/lib/main.c	2012-10-11 14:19:12.699247000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/TestFramework/lib/main.c	2012-10-11 14:22:17.487226000 -0400
@@ -37,7 +37,7 @@
     int retval=0;
 
     // Trap the interupt signal
-    signal(SIGINT, s_SignalHandler);
+    /* signal(SIGINT, s_SignalHandler); keep a Ctrl-C working */
 
     // No buffer is used for quick output
     setvbuf(stdout, NULL, _IONBF, 0);
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/dirs SSPK-release-V1_5/Source/Platform/LinuxSample/dirs
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/dirs	2012-10-11 14:19:12.387383000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/dirs	2012-10-11 14:22:17.489222000 -0400
@@ -4,3 +4,4 @@
   HALimpl \
   PALimpl \
   TestFramework \
+
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/inc/platPrivate.h SSPK-release-V1_5/Source/Platform/LinuxSample/inc/platPrivate.h
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/inc/platPrivate.h	2012-10-11 14:19:12.571375000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/inc/platPrivate.h	2012-10-11 14:22:17.490227000 -0400
diff -ru 20130205_V1.5Plus_mediaCL69986.orig/Source/Platform/Nexus/inc/platPrivate.h 20130205_V1.5Plus_mediaCL69986/Source/Platform/Nexus/inc/platPrivate.h
@@ -29,6 +29,8 @@
 
 extern void SetThreadLastSocketError_priv(pkRESULT lastError);
 extern pkRESULT GetThreadLastSocketError_priv(void);
+struct IPTV_HAL_Decoder_Resources;
+extern int IPTV_HAL_Decoder_HALInit_priv(const struct IPTV_HAL_Decoder_Resources *resources);
 
 extern int TF_Main(int argc, char *argv[]);
 
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/inc/platSockets.h SSPK-release-V1_5/Source/Platform/LinuxSample/inc/platSockets.h
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/inc/platSockets.h	2012-10-11 14:19:12.580367000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/inc/platSockets.h	2012-10-11 14:22:17.492225000 -0400
@@ -142,8 +142,8 @@
 
 
 /* Socket Types */
-#define SOCKET_STREAM     1               /* stream socket */
-#define SOCKET_DGRAM      2               /* datagram socket */
+#define SOCKET_STREAM     2               /* stream socket */
+#define SOCKET_DGRAM      1               /* datagram socket */
 #define SOCKET_RAW        3               /* raw-protocol interface */
 
 /* inet addr constants */
diff -ru SSPK-release-V1_5.orig/Source/Platform/LinuxSample/inc/wintypes.h SSPK-release-V1_5/Source/Platform/LinuxSample/inc/wintypes.h
--- SSPK-release-V1_5.orig/Source/Platform/LinuxSample/inc/wintypes.h	2012-10-11 14:19:12.609340000 -0400
+++ SSPK-release-V1_5/Source/Platform/LinuxSample/inc/wintypes.h	2012-10-11 14:22:17.494221000 -0400
@@ -189,13 +189,13 @@
 #endif
 
 #ifndef __cdecl
-#define __cdecl __attribute((stdcall))
+#define __cdecl // __attribute((stdcall))
 #endif
 
-#define API                 __attribute((stdcall))
-#define STDAPICALLTYPE      __attribute((stdcall))
-#define STDMETHODCALLTYPE   __attribute((stdcall))
-#define CALLBACK            __attribute((stdcall))
+#define API                 // __attribute((stdcall))
+#define STDAPICALLTYPE      // __attribute((stdcall))
+#define STDMETHODCALLTYPE   // __attribute((stdcall))
+#define CALLBACK            // __attribute((stdcall))
 #define DECLSPEC_SELECTANY
 #define DECLSPEC_NOVTABLE
 
@@ -354,7 +354,7 @@
 // Standard Interface declarations
 //----------------------------------------------
 #define interface struct
-#define WINAPI  __attribute((stdcall))
+#define WINAPI  // __attribute((stdcall))
 
 #define INVALID_HANDLE_VALUE ((HANDLE)(LONG_PTR)-1)
 #define INVALID_FILE_SIZE ((DWORD)0xFFFFFFFF)
diff -ru SSPK-release-V1_5.orig/Source/Test/Apps/Integration/dirs SSPK-release-V1_5/Source/Test/Apps/Integration/dirs
--- SSPK-release-V1_5.orig/Source/Test/Apps/Integration/dirs	2012-10-11 14:19:12.748198000 -0400
+++ SSPK-release-V1_5/Source/Test/Apps/Integration/dirs	2012-10-11 14:22:17.495228000 -0400
@@ -2,4 +2,4 @@
   SmoothTestLib	\
   SSPKIntegrationTests	\
   SSPKStressTests \
-  SSPKHeuristicsTests \
\ No newline at end of file
+  SSPKHeuristicsTests \
