diff -BNru bsa0107_00.12.00-netapp-v1_org/3rdparty/embedded/brcm/bsa/udrv/linux/client/uipc_cl_sco.c bsa0107_00.12.00-netapp-v1/3rdparty/embedded/brcm/bsa/udrv/linux/client/uipc_cl_sco.c
--- bsa0107_00.12.00-netapp-v1_org/3rdparty/embedded/brcm/bsa/udrv/linux/client/uipc_cl_sco.c	2015-10-14 20:47:44.000000000 -0400
+++ bsa0107_00.12.00-netapp-v1/3rdparty/embedded/brcm/bsa/udrv/linux/client/uipc_cl_sco.c	2016-01-11 15:53:30.039045376 -0500
@@ -233,7 +233,7 @@
     {
         uipc_cl_sco_cb.p_cback = NULL;
     }
-
+    uipc_thread_stop(uipc_cl_sco_cb.thread);
 }
 
 /*******************************************************************************
diff -BNru bsa0107_00.12.00-netapp-v1_org/3rdparty/embedded/brcm/bsa/udrv/linux/uipc_rb.c bsa0107_00.12.00-netapp-v1/3rdparty/embedded/brcm/bsa/udrv/linux/uipc_rb.c
--- bsa0107_00.12.00-netapp-v1_org/3rdparty/embedded/brcm/bsa/udrv/linux/uipc_rb.c	2015-10-14 20:47:44.000000000 -0400
+++ bsa0107_00.12.00-netapp-v1/3rdparty/embedded/brcm/bsa/udrv/linux/uipc_rb.c	2016-01-11 15:53:30.039045376 -0500
@@ -790,7 +790,11 @@
         APPL_TRACE_ERROR0("ERROR uipc_rb_select: uipc_sema_lock (rd_lockid) fail");
         return -1;
     }
-
+    if(p_uipc_rb->p_hdr == NULL)
+    {
+        printf("Debug check p_hdr == NULL");
+        return -1;
+    }
 
     return (p_hdr->w_idx + p_uipc_rb->size - p_hdr->r_idx) % p_uipc_rb->size;
 }
diff -BNru bsa0107_00.12.00-netapp-v1_org/3rdparty/embedded/bsa_examples/linux/server/build/bsa_server.txt bsa0107_00.12.00-netapp-v1/3rdparty/embedded/bsa_examples/linux/server/build/bsa_server.txt
--- bsa0107_00.12.00-netapp-v1_org/3rdparty/embedded/bsa_examples/linux/server/build/bsa_server.txt	2015-10-23 15:16:24.000000000 -0400
+++ bsa0107_00.12.00-netapp-v1/3rdparty/embedded/bsa_examples/linux/server/build/bsa_server.txt	2016-01-11 15:54:13.367046858 -0500
@@ -38,7 +38,7 @@
 # 2- add a BSA server command line option to change the default baudrate to the specified
 #    baudrate.  This allows terminating and restarting bsa_server without having to
 #    disconnect the BT controller
-BSA_CHANGE_STARTUP_BAUDRATE = USERIAL_BAUD_2_5M
+BSA_CHANGE_STARTUP_BAUDRATE = USERIAL_BAUD_1M
 
 # include BTSNOOP file capture
 BTSNOOPDISP_INCLUDED = TRUE
