--- Makefile	2015-06-25 12:33:10.224960644 -0700
+++ Makefile.blive	2015-06-25 11:39:14.726111884 -0700
@@ -127,7 +127,7 @@
 $(LIVEMEDIA_LIB): $(LIVEMEDIA_LIB_OBJS) \
     $(PLATFORM_SPECIFIC_LIB_OBJS)
 	$(LIBRARY_LINK)$@ $(LIBRARY_LINK_OPTS) \
-		$(LIVEMEDIA_LIB_OBJS)
+		$(LIVEMEDIA_LIB_OBJS) ../../../blive_ext/lib/arm-linux.debug/lm*.o
 
 Media.$(CPP):		include/Media.hh
 include/Media.hh:	include/liveMedia_version.hh
@@ -430,8 +430,8 @@
 install: install1 $(INSTALL2)
 install1: $(LIVEMEDIA_LIB)
 	 install -d $(DESTDIR)$(PREFIX)/include/liveMedia $(DESTDIR)$(LIBDIR)
-	 install -m 644 include/*.hh $(DESTDIR)$(PREFIX)/include/liveMedia
-	 install -m 644 $(LIVEMEDIA_LIB) $(DESTDIR)$(LIBDIR)
+	 install -C -m 644 include/*.hh $(DESTDIR)$(PREFIX)/include/liveMedia
+	 install -C -m 644 $(LIVEMEDIA_LIB) $(DESTDIR)$(LIBDIR)
 install_shared_libraries: $(LIVEMEDIA_LIB)
 	 ln -fs $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).$(SHORT_LIB_SUFFIX)
 	 ln -fs $(NAME).$(LIB_SUFFIX) $(DESTDIR)$(LIBDIR)/$(NAME).so
